examples/diffdope_ros/scripts/server.py

```diff
@@ -0,0 +1,342 @@
+        except Exception as e:
+            rospy.logerr(f"CvBridge Error: {e}")
+
+        rgb_frame = self.__resize_image(rgb_frame)
+        rgb_tensor = torch.tensor(rgb_frame).float()
+        rgb_image = dd.Image(img_tensor=rgb_tensor, img_resize=self.cfg.image_resize)
+        return rgb_image
+
+    def __convert_depth_frame_to_diffdope_image(self, cv_bridge, depth_frame):
+        depth_frame = cv_bridge.imgmsg_to_cv2(depth_frame, depth_frame.encoding)
+        depth_frame = depth_frame.astype(np.int16)
+        depth_frame = cv2.flip(depth_frame, 0)  # needed by diff-dope
+        depth_frame = self.__resize_image(depth_frame, is_depth=True)
+        depth_tensor = torch.tensor(depth_frame).float()
+        depth_image = dd.Image(
+            img_tensor=depth_tensor, img_resize=self.cfg.image_resize, depth=True
+        )
+
+        return depth_image
+
+    def __convert_segmentation_frame_to_diffdope_image(
+        self, cv_bridge, rgb_frame, point_in_camera
+    ):
+        try:
+            cv_image = cv_bridge.imgmsg_to_cv2(rgb_frame, "bgr8")
+            rgb_frame = cv2.cvtColor(cv_image, cv2.COLOR_BGR2RGB)
+
+            segmentation_frame = self.segment_anything.segment(
+                rgb_frame, point_in_camera
+            )
+        except Exception as e:
+            rospy.logerr(f"CvBridge Error: {e}")
+
+        # Make segmentation a 3 channel RGB image.
+        segmentation_frame = cv2.merge(
+            (segmentation_frame, segmentation_frame, segmentation_frame)
+        )
+        segmentation_frame = segmentation_frame / 255.0
+        segmentation_frame = cv2.flip(segmentation_frame, 0)  # needed by diff-dope
+
+        segmentation_frame = self.__resize_image(segmentation_frame)
+        segmentation_tensor = torch.tensor(segmentation_frame).float()
+
+        segmentation_image = dd.Image(
+            img_tensor=segmentation_tensor, img_resize=self.cfg.image_resize
+        )
+
+        return segmentation_image
+
+    def __generate_scene(self, rgb_frame, depth_frame, point_in_camera):
+        cv_bridge = CvBridge()
+        rgb_image = self.__convert_rgb_frame_to_diffdope_image(cv_bridge, rgb_frame)
+        depth_image = self.__convert_depth_frame_to_diffdope_image(
+            cv_bridge, depth_frame
+        )
+        segmentation_image = self.__convert_segmentation_frame_to_diffdope_image(
+            cv_bridge, rgb_frame, point_in_camera
+        )
+
+        scene = dd.Scene(
+            tensor_rgb=rgb_image,
+            tensor_depth=depth_image,
+            tensor_segmentation=segmentation_image,
+            image_resize=self.cfg.image_resize,
+        )
+
+        return scene
+
+    def __generate_3d_object(self, goal: TargetObject):
+        dope_pose = goal.pose.pose
+        dope_position = dope_pose.position
+        position = np.array([dope_position.x, dope_position.y, dope_position.z])
+
+        # Convert to mm
+        position *= 1_000
```

Comment:
From your provided example, it seems that Diff-DOPE expects the input position in millimetres, here I convert DOPE's reading from metres to millimetres if my assumption is correct.

Reply:
:D